Compute best practices updates (#1088)
microsoft/finops-toolkit: docs/_docs/best-practices/compute/compute.md

<details open markdown="1">
   <summary class="fs-2 text-uppercase">On this page</summary>

- [Advisor](#advisor)
- [Virtual machines](#virtual-machines)
- [App Service plans](#app-service-plans)
- [Azure Kubernetes Service](#azure-kubernetes-service)
- [🙋‍♀️ Looking for more?](#️-looking-for-more)
- [🧰 Related tools](#-related-tools)

</details>

---

## Advisor

### List of cost recommendations for Compute

This Azure Resource Graph (ARG) query retrieves a list of Azure Advisor recommendations specifically for compute resources. It filters the recommendations to include only those related to virtual machines, scale sets, and other compute services, providing insights into potential cost savings.

<h4>Category</h4>

Cost optimization

<h4>Query</h4>

> ```kql

Reviewer: nit: We should establish some KQL language standards. If you happen to know of any, we should document them in the wiki. No worries if you haven't.

advisorresources
| where type == "microsoft.advisor/recommendations"

> | where tostring (properties.category) has "Cost"

Reviewer: ```suggestion
| where tostring(properties.category) has "Cost"
```

| where properties.shortDescription.problem has "underutilized"

> | where properties.impactedField has "Compute" or properties.impactedField has "Container" or properties.impactedField has "Web"

Reviewer: Move each `or` to a new line so it's more readable online. Web should be in the Web category and not in Compute according to FOCUS service categories.
```suggestion
| where properties.impactedField has "Compute"
    or properties.impactedField has "Container"
```

> | project AffectedResource=tostring(properties.resourceMetadata.resourceId),Impact=properties.impact,resourceGroup,AdditionaInfo=properties.extendedProperties,subscriptionId,Recommendation=tostring(properties.shortDescription.problem)

Reviewer: There should be spaces around `=`, after commas, and long lines should be broken up on multiple lines.

| project id, PowerState, VMLocation, resourceGroup, subscriptionId
```

<br>

> ### Query: List Virtual Machines deallocated

Reviewer: ```suggestion
### Query: List deallocated virtual machines
```

This Azure Resource Graph (ARG) query identifies Virtual Machines (VMs) in your Azure environment that are in the 'deallocated' state. It retrieves details about their power state, location, resource group, and subscription ID.

<h4>Category</h4>

Waste reduction

<h4>Query</h4>

```kql
resources
| where type =~ 'microsoft.compute/virtualmachines'
    and tostring(properties.extended.instanceView.powerState.displayStatus) == 'VM deallocated'

> | where resourceGroup in ({ResourceGroup})

Reviewer: What is `{ResourceGroup}` here?

Reviewer 2: This is used by the workbook to filter the results based on a parameter called ResourceGroup. We can remove it from this query.

> | extend  PowerState=tostring(properties.extended.instanceView.powerState.displayStatus), VMLocation=location, resourceGroup=strcat('/subscriptions/',subscriptionId,'/resourceGroups/',resourceGroup)

Reviewer: We should probably standardize on only one property per extend for readability.
```suggestion
| extend  PowerState = tostring(properties.extended.instanceView.powerState.displayStatus)
| extend  VMLocation = location
| extend  resourceGroup = strcat('/subscriptions/', subscriptionId, '/resourceGroups/', resourceGroup)
```

| order by id asc
| project id, PowerState, VMLocation, resourceGroup, subscriptionId
```

<br>

> ### Query: List of Virtual Machines with their associated disks

Reviewer: ```suggestion
### Query: List of virtual machines with their associated disks
```

This Resource Graph query retrieves a comprehensive list of all Virtual Machines (VMs) in your Azure environment, along with details of their associated disks. It provides insights into the storage configuration of each VM, helping you manage and optimize your storage resources effectively.

<h4>Category</h4>

Resource management

<h4>Query</h4>

```kql

> Resources | where type == "microsoft.compute/virtualmachines"
> | extend osDiskId= tostring(properties.storageProfile.osDisk.managedDisk.id)
>         | join kind=leftouter(resources
>             | where type =~ 'microsoft.compute/disks'
>             | where properties !has 'Unattached'
>             | where properties has 'osType'
>             | project OS = tostring(properties.osType), osSku = tostring(sku.name), osDiskSizeGB = toint(properties.diskSizeGB), osDiskId=tostring(id)) on osDiskId
>         | join kind=leftouter(Resources
>             | where type =~ 'microsoft.compute/disks'
>             | where properties !has "osType"
>             | where properties !has 'Unattached'
>             | project sku = tostring(sku.name), diskSizeGB = toint(properties.diskSizeGB), id = managedBy
>             | summarize sum(diskSizeGB), count(sku) by id, sku) on id
> | project vmId=id, subscriptionId, resourceGroup, OS, location, osDiskId, osSku, osDiskSizeGB, DataDisksGB=sum_diskSizeGB, diskSkuCount=count_sku
> | sort by diskSkuCount desc

Reviewer: 1. Make sure the queries are formatted correctly.
2. You're using a case-insensitive type filter on some lines and case-sensitive on the first where. Which is correct?
3. Let's always use single instead of double quotes.
4. Make sure table names are cased consistently. You have "resources" and "Resources".
5. Doesn't KQL use `order by` instead of `sort by`?
```suggestion
Resources
| where type == 'microsoft.compute/virtualmachines'
| extend osDiskId= tostring(properties.storageProfile.osDisk.managedDisk.id)
| join kind=leftouter(
    Resources
    | where type =~ 'microsoft.compute/disks'
    | where properties !has 'Unattached'
    | where properties has 'osType'
    | project
        OS = tostring(properties.osType),
        osSku = tostring(sku.name),
        osDiskSizeGB = toint(properties.diskSizeGB),
         osDiskId=tostring(id)
) on osDiskId
| join kind=leftouter(
    Resources
    | where type =~ 'microsoft.compute/disks'
    | where properties !has "osType"
    | where properties !has 'Unattached'
    | project
        sku = tostring(sku.name),
        diskSizeGB = toint(properties.diskSizeGB),
        id = managedBy
    | summarize sum(diskSizeGB), count(sku) by id, sku
) on id
| project
    vmId = id,
    subscriptionId,
    resourceGroup,
    OS,
    location,
    osDiskId,
    osSku,
    osDiskSizeGB,
    DataDisksGB = sum_diskSizeGB,
    diskSkuCount = count_sku
| order by diskSkuCount desc
```

<br>

> ### Query: Virtual machine scale set details

Reviewer: Shouldn't this be in the VMSS section? (Unless I missed it.)

This query analyzes Virtual Machine Scale Sets (VMSS) in your Azure environment based on their SKU, spot VM priority, and priority mix policy. It provides insights for cost optimization and resource management strategies.

<h4>Category</h4>

Resource management

<h4>Query</h4>

```kql
resources
| where type =~ 'microsoft.compute/virtualmachinescalesets'
| extend SpotVMs = tostring(properties.virtualMachineProfile.priority)
| extend SpotPriorityMix = tostring(properties.priorityMixPolicy)
| extend SKU = tostring(sku.name)
| extend resourceGroup = strcat('/subscriptions/', subscriptionId, '/resourceGroups/', resourceGroup)
| project id, SKU, SpotVMs, SpotPriorityMix, subscriptionId, resourceGroup, location
```

<br>

> ## App Service plans

Reviewer: Move this to the Web category.

### Query: App Service plans with no hosted applications

This Resource Graph query identifies all App Service plans in your Azure environment that do not have any hosted applications. It helps you pinpoint unused resources, enabling you to optimize costs and improve resource management.

<h4>Category</h4>

Waste management

<h4>Query</h4>

```kql
resources
| where type =~ "microsoft.web/serverfarms"
| where properties.numberOfSites == 0
| extend Details = pack_all()
| project Resource=id, resourceGroup, location, subscriptionId, Sku=sku.name, Tier=sku.tier, tags ,Details
```

<br>

> ## Azure Kubernetes Service

Reviewer: Move this back to the top. Services should be sorted alphabetically.
